### PR TITLE
#8yVIZ5wE Driver and operator able to update their profile

### LINKED ISSUE #24: Heroku URI
### Description
Your README should have a link to your production app.

### PR DESCRIPTION
**Pull request by Niyiragira**
###  What does this PR do?
This PR requests to merge ft-update-profile-zqKf1dcP into develop

###  Description of Task to be completed?
The operator and driver should be able to update information about their own profile.
The operator and driver profile should be able accessible publicly.
Appropriate error messages are displayed on unsuccessful attempts.

###  How should this be manually tested?
1. You can use this app https://phantom-caba-ft-updates-xy8zlh.herokuapp.com/api-docs to send below requests

2. Or you can use Postman under this link : https://phantom-caba-ft-updates-xy8zlh.herokuapp.com
 
3. And then login (POST /api/v1/auth/login) using these credentials to have permission :email: **admin@phantom.com** and password:**admin** 

4. create user (POST /api/v1/auth/register)

5. Login (POST /api/v1/auth/login) using these to have permission :email: **<<email you registered>>** and password:**<<password you received>>* 

6. update yourself (PATCH /api/v1/auth/updateProfile)  .
      Copy  token provided with login and paste in headers under key of Authorization.

7. GET one user using his/her id at /api-docs/api/v1/auth/profiles/:id

###  What are the relevant pivotal tracker stories?
[zqKf1dcP](https://trello.com/c/zqKf1dcP/38-update-profile)

###  Screenshots (if appropriate)
Screen Shot 2020-10-23 at 08 19 42

### CODE REVIEW

File: src/test/testProfileUpdate.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      .patch("/api/v1/auth/updateProfile")
+      .set("Authorization", updateMock.token.oparator)
+      .send(registerMock.nameIsEmpty)
+      .end((err,res)=>{
+        expect(res).to.have.status(400);
+        done();
+      })
+  }); 
+  });
+
```
Comment (Nkbtemmy) [inline src/test/testProfileUpdate.js]: Request new line

**Review by Nkbtemmy — CHANGES_REQUESTED** [review]

File: src/test/testProfileUpdate.js (hunk 1 above)
Comment (Niyiragira) [reply src/test/testProfileUpdate.js]: You can see it's there. Any other way to add it maybe?

**Review by Nkbtemmy — APPROVED** [review]

**Review by karigirwa — APPROVED** [review]

**Review by ceelogre — CHANGES_REQUESTED** [review]
The description of your pr is not complete, why do we have to use specific login credentials (…gerard@gmail) 

When you say to update yourself what do you mean?

When you say "token" do you mean we should use "token" key in the auth headers?

finally, get profile by id doesn’t seem to work, refer to the screenshots below.

Capture d’écran 2020-11-03 à 11 01 32


Capture d’écran 2020-11-03 à 11 00 47
Capture d’écran 2020-11-03 à 11 02 31

**Review by blaise82 — DISMISSED** [review]
either you can move to update the email or you can put away that when an email is updated it's also validated

File: src/controller/userControllers.js [hunk 2]
```diff
@@ -183,6 +177,53 @@ class userController {
...
+        message:res.__('User with the specified id is not found.')
+      })
+    } catch (error) {
+      return res.status(500).send(error.message);
+    }
+  }
+
+ static async updateUserSelf(req, res){
+    try {
+      const email=req.user.email
```
Comment (blaise82) [inline src/controller/userControllers.js]: user destruction

File: src/controller/userControllers.js [hunk 3]
```diff
@@ -183,6 +177,53 @@ class userController {
 
         }
    }
+
+   static async getUserById(req, res){
+    try {
+      const { userId } = req.params;
+      const user = await Model.User.findOne({
```
Comment (blaise82) [inline src/controller/userControllers.js]: make the User Model a global variable and use it without always putting the Model 
ex: const { User } = Model

**Review by blaise82 — COMMENTED** [review]
validate gender update (allow only valid genders to be put in)

Screen Shot 2020-11-03 at 11 34 10

**Review by ncutixavier — APPROVED** [review]

File: src/controller/userControllers.js [hunk 4]
```diff
@@ -143,17 +145,11 @@ class userController {
...
         
         const url  = req.params;
-        console.log('url=',url)
         const token = url.token
-        console.log('token=',token)
         const payload = decode(token);
-
-        
-        console.log('payload=',payload)
         const email  = payload.email;
```
Comment (ceelogre) [inline src/controller/userControllers.js]: Use destructuring too.

File: src/test/testProfileUpdate.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  }); 
+  it("Should not update user if name is empty ", done=>{
+    chai 
+      .request(app)
+      .patch("/api/v1/auth/updateProfile")
+      .set("Authorization", updateMock.token.oparator)
+      .send(registerMock.invalidEmail)
+      .end((err,res)=>{
+        expect(res).to.have.status(401);
+        done();
```
Comment (ceelogre) [inline src/test/testProfileUpdate.js]: Add assertion for messages or datatype after status.

**Review by ceelogre — CHANGES_REQUESTED** [review]
Look at the return messages on unsuccessful attempts and make them match the exact exceptions that happened.

**Review by ceelogre — APPROVED** [review]

**Pull request merged**